inc/Engine/CriticalPath/CriticalCSSSubscriber.php

```diff
@@ -687,39 +594,82 @@ public function async_css( $buffer ) {
 		 */
 		$css_pattern = apply_filters(
 			'rocket_async_css_regex_pattern',
-			'/(?=<link[^>]*\s(rel\s*=\s*[\'"]stylesheet["\']))<link[^>]*\shref\s*=\s*[\'"]([^\'"]+)[\'"](.*)>/iU'
+			'/(?=<link[^>]*\s(rel\s*=\s*[\'"]stylesheet["\']))<link(?=[^<>]*?(?:(?<type>type\s*=\s*[\'"][^<>"\']*[\'"])|>))(?=[^<>]*?(?:(?<onload>onload\s*=\s*(.*))|>))(?=[^<>]*?(?:(?<media>media\s*=\s*[\'"][^<>"\']*[\'"])|>))(?=[^<>]*?(?:href\s*=\s*[\'"](?<url>.*?)[\'"]|>))(?=[^<>]*?(?:(?<rel>rel\s*=\s*[\'"]>[^<>"\']*)[\'"]|>))(?:.*?<\/\1>|[^<>]*>)/ix'
 		);
 
 		// Get all css files with this regex.
 		preg_match_all( $css_pattern, $buffer, $tags_match );
+
 		if ( ! isset( $tags_match[0] ) ) {
 			return $buffer;
 		}
 
-		$noscripts = '';
+		return $this->async_css_tag_update( $tags_match, $buffer );
+	}
 
+	/**
+	 * Updates the HTML code by defering all CSS matches files.
+	 *
+	 * @param  array  $tags_match All matched css files.
+	 * @param  string $buffer     HTML code.
+	 * @return string             Modified HTML code with defer CSS.
+	 */
+	private function async_css_tag_update( $tags_match, $buffer ) {
+		$excluded_css = array_flip( $this->critical_css->get_exclude_async_css() );
+		$noscripts    = '';
 		foreach ( $tags_match[0] as $i => $tag ) {
 			// Strip query args.
-			$path = rocket_extract_url_component( $tags_match[2][ $i ], PHP_URL_PATH );
+			$path = rocket_extract_url_component( $tags_match['url'][ $i ], PHP_URL_PATH );
 
 			// Check if this file should be deferred.
 			if ( isset( $excluded_css[ $path ] ) ) {
 				continue;
 			}
 
-			$preload = str_replace( 'stylesheet', 'preload', $tags_match[1][ $i ] );
-			$onload  = preg_replace( '~' . preg_quote( $tags_match[3][ $i ], '~' ) . '~iU', ' as="style" onload=""' . $tags_match[3][ $i ] . '>', $tags_match[3][ $i ] );
-			$tag     = str_replace( $tags_match[3][ $i ] . '>', $onload, $tag );
-			$tag     = str_replace( $tags_match[1][ $i ], $preload, $tag );
-			$tag     = str_replace( 'onload=""', 'onload="this.onload=null;this.rel=\'stylesheet\'"', $tag );
-			$buffer  = str_replace( $tags_match[0][ $i ], $tag, $buffer );
+			$original_media = 'all';
+			$tag            = str_replace( $tags_match['type'][ $i ], ' as="style" ' . $tags_match['type'][ $i ] . ( empty( $tags_match['media'][ $i ] ) ? ' media="print"' : '' ) . ( empty( $tags_match['onload'][ $i ] ) ? ' onload=""' : '' ), $tag );
+			if ( ! empty( $tags_match['media'][ $i ] ) ) {
+				preg_match( '/media\s*=\s*[\'"](?<media>.*)[\'"]/ix', $tags_match['media'][ $i ], $media_match );
+				$original_media = $media_match['media'];
+				$tag            = str_replace( $tags_match['media'][ $i ], 'media="print"', $tag );
```

Comment:
To help with readability and multiple lookups, let's store in a variable:

```suggestion
				$media_attr = $tags_match['media'][ $i ];
				preg_match( '/media\s*=\s*[\'"](?<media>.*)[\'"]/ix', $media_attr, $media_match );
				$original_media = $media_match['media'];
				$tag            = str_replace( $media_attr, 'media="print"', $tag );
```